File: ZooJava/src/main/java/com/example/zoojava/utils/DataSystem.java
package com.example.zoojava.utils;

import com.example.zoojava.repositories.animals.AnimalsRepository;
import com.example.zoojava.repositories.animals.AnimalsRepositoryImpl;
import com.example.zoojava.repositories.employees.EmployeesRepository;
import com.example.zoojava.repositories.employees.EmployeesRepositoryImpl;
import com.example.zoojava.utils.csv.ImportAnimalCsv;
import com.example.zoojava.utils.csv.ImportAnimalCsvImpl;
import com.example.zoojava.utils.csv.ImportEmployeeCsv;
import com.example.zoojava.utils.csv.ImportEmployeeCsvImpl;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.sql.SQLException;

/**
 * Clase para el manejo de daots tanto como añadirlos o la limpieza de estos
 */
public class DataSystem {
    private final ImportAnimalCsv  animalCsv= new ImportAnimalCsvImpl();
    private final ImportEmployeeCsv employeeCsv= new ImportEmployeeCsvImpl();
    private final AnimalsRepository animalsRepository= AnimalsRepositoryImpl.getInstance();
    private final EmployeesRepository employeesRepository= EmployeesRepositoryImpl.getInstance();


    
    /**
     * Añadir los datos necesarios.
     */
    public void addData(){
        addAnimalData();
        addEmployeeData();
    }


    /**
     * Añadir los datos de los empleados
     */
    private void addEmployeeData() {
        var listEmployee = employeeCsv.importData();
        listEmployee.forEach((a) -> {
            try {
                employeesRepository.add(a);
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        });
    }


    /**
     * Añadir los datos de los animales
     */
    private void addAnimalData() {
        var listAnimals = animalCsv.importData();


        listAnimals.forEach((a) -> {
            try {
                animalsRepository.add(a);
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        });
    }


    /**
     * Backup de los animales y empleados.
     * Limpieza de base de datos.
     * Al cerrar el programa.
     */
    public void backupAndClean() {
        backupAnimals();
        backupEmployees();
        cleanDataBase();
    }


    /**
     * Limpieza base de datos al cerrar el programa.
     */
    private void cleanDataBase() {
        try {
            animalsRepository.removeAll();
            employeesRepository.removeAll();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }


    /**
     * Backup de los empleados
     */
    private void backupEmployees() {
        try (PrintWriter pw = new PrintWriter(new FileWriter(Globals.EMPLOYEE_CSV, false))) {
            var employees = employeesRepository.findAll();
            pw.println("name;email;password;birthdate;isAdmin");
            employees.forEach((a) -> pw.println(a.toCsv()));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }


    /**
     * Para hacer backup de los animales al terminar de utilizar el programa.
     */
    private void backupAnimals() {
        try (PrintWriter pw = new PrintWriter(new FileWriter(Globals.ANIMALS_CSV, false))) {
            var animals = animalsRepository.findAll();
            pw.println("id;name;type;birthdate;img");
            animals.forEach((a) -> pw.println(a.toCsv()));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

}

File: ZooKotlin/src/main/kotlin/com/example/zookotlin/utils/DataSystem.kt
package com.example.zookotlin.utils

import com.example.zookotlin.repositories.animals.AnimalsRepository
import com.example.zookotlin.repositories.animals.AnimalsRepositoryImpl
import com.example.zookotlin.repositories.employees.EmployeesRepository
import com.example.zookotlin.repositories.employees.EmployeesRepositoryImpl
import com.example.zookotlin.utils.csv.ImportAnimalCsv
import com.example.zookotlin.utils.csv.ImportAnimalCsvImpl
import com.example.zookotlin.utils.csv.ImportEmployeeCsv
import com.example.zookotlin.utils.csv.ImportEmployeeCsvImpl
import java.io.FileWriter
import java.io.PrintWriter

/**
 * Clase para el manejo de daots tanto como añadirlos o la limpieza de estos
 */
class DataSystem {
    private var animalCsv: ImportAnimalCsv = ImportAnimalCsvImpl()
    private var employeeCsv: ImportEmployeeCsv = ImportEmployeeCsvImpl()
    private var animalsRepository: AnimalsRepository = AnimalsRepositoryImpl.getInstance()!!
    private var employeeRepository: EmployeesRepository = EmployeesRepositoryImpl.getInstance()!!


    /**
     * Añadir los datos necesarios
     */
    fun addData(){
        addAnimalData()
        addEmployeeData()
    }


    /**
     * Añadir los datos de los empleados
     */
    private fun addEmployeeData() {
        val listEmployees = employeeCsv.importData()
        listEmployees.forEach {a -> employeeRepository.add(a)}
    }


    /**
     * Añadir los datos de los animales
     */
    private fun addAnimalData() {
        val listAnimals = animalCsv.importData()
        listAnimals.forEach {a -> animalsRepository.add(a) }
    }


    /**
     * Backup de los aniumales y empleados.
     * Limpieza de la base de datos.
     * Al cerrar el programa
     */
    fun backupAndClean(){
        backupAnimals()
        backupEmployees()
        cleanDataBase()
    }


    /**
     * Limpieza de la base de datos.
     */
    private fun cleanDataBase() {
        animalsRepository.removeAll()
        employeeRepository.removeAll()
    }

    /**
     * Backup de los empleados en CSV.
     */
    private fun backupEmployees() {
        val pw = PrintWriter(FileWriter(Globals.EMPLOYEE_CSV,false))
        val employees = employeeRepository.findAll()
        pw.println("name;email;password;birthdate;isAdmin")
        employees.forEach {a -> pw.println(a.toCsv())}
    }


    /**
     * Backup de los animales en CSV.
     */
    private fun backupAnimals() {
        val pw = PrintWriter(FileWriter(Globals.ANIMAL_CSV,false))
        val animals = animalsRepository.findAll()
        pw.println("id;name;type;birthdate;img")
        animals.forEach {a -> pw.println(a.toCsv())}
    }
}